### PR TITLE
Ensure fed-sender catchup does not block for full state

### LINKED ISSUE #15220: Federation catchup for a destination can block indefinitely on un-partial stating
**What we think happened**
1. A user faster joined a remote room
2. Some events were sent into the room, but not transmitted successfully
3. The user left the room before it became full state
4. The federation catchup loop tried to transmit events in the room, but blocked on full state. This also blocked \*all\* catchup transmission for the destination.

**Stack trace**
```
  File "/home/synapse/src/synapse/metrics/background_process_metrics.py", line 242, in run
    return await func(*args, **kwargs)
  File "/home/synapse/src/synapse/federation/sender/per_destination_queue.py", line 321, in _transaction_transmission_loop
    await self._catch_up_transmission_loop()
  File "/home/synapse/src/synapse/federation/sender/per_destination_queue.py", line 544, in _catch_up_transmission_loop
    new_pdus = await filter_events_for_server(
  File "/home/synapse/src/synapse/visibility.py", line 650, in filter_events_for_server
    event_to_memberships = await _event_to_memberships(
  File "/home/synapse/src/synapse/visibility.py", line 746, in _event_to_memberships
    event_to_state_ids = await storage.state.get_state_ids_for_events(
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/storage/controllers/state.py", line 267, in get_state_ids_for_events
    event_to_groups = await self.get_state_group_for_events(
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/storage/controllers/state.py", line 374, in get_state_group_for_events
    await self._partial_state_events_tracker.await_full_state(event_ids)
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/storage/util/partial_state_events_tracker.py", line 82, in await_full_state
    logger.info(
```

### PR DESCRIPTION
**Pull request by DMRobertson**
Fixes #15220.

Previous iterations were https://github.com/matrix-org/synapse/pull/15228 https://github.com/matrix-org/synapse/pull/15241.

### CODE REVIEW

**Comment by DMRobertson** [comment]
We should get this into the release branch, do an RC on Monday.

Need to revert https://github.com/matrix-org/synapse/commit/ab629c17ccfbc98e8297e7fd3fcf03b641bd3cc0on hotfixes once this lands too.

File: synapse/federation/sender/per_destination_queue.py [hunk 1]
```diff
@@ -516,6 +516,11 @@ async def _catch_up_transmission_loop(self) -> None:
                     # If the event is in the extremities, then great! We can just
                     # use that without having to do further checks.
                     room_catchup_pdus = [pdu]
+                elif self._store.is_partial_state_room(pdu.room_id):
+                    # We can't be sure which events the destination should
+                    # see using only partial state. Avoid doing so, and just retry
+                    # sending our the oldest PDU the remote is missing from us.
+                    room_catchup_pdus = [pdu]
```
Comment (squahtx) [inline synapse/federation/sender/per_destination_queue.py:523]: I think this makes sense after @reivilibre's explanation (modulo confusion about oldest vs newest).
The key bit I was missing before was that we're not filtering the catchup PDUs (which are okay to send out), but the forward extremities (which may not be okay to send out).

**Review by squahtx — COMMENTED** [review]
Looks good!

I'm still confused about whether catchup PDUs are the oldest or newest PDUs in each room which we wanted to send out.

Many thanks for dealing with this. It's thoroughly confusing.

File: synapse/federation/sender/per_destination_queue.py [hunk 2]
```diff
@@ -516,6 +516,11 @@ async def _catch_up_transmission_loop(self) -> None:
                     # If the event is in the extremities, then great! We can just
                     # use that without having to do further checks.
                     room_catchup_pdus = [pdu]
+                elif self._store.is_partial_state_room(pdu.room_id):
```
Comment (DMRobertson) [inline synapse/federation/sender/per_destination_queue.py]: https://github.com/matrix-org/synapse/pull/14519 please

**Review by squahtx — APPROVED** [review]
Looks good. Thank you for taking this on!

**Comment by DMRobertson** [comment]
AFAICS the test failures are flakes.

**Pull request merged**